Stop keypresses being caught by other elements when they happen in a CustomSelectControl

opr opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

## Description
<!-- Please describe what you have changed or added -->
In the `<ul>` output by `CustomSelectControl` I have added a function to handle `onKeyDown` - this stops the event from propagating, but still calls the `onKeyDown` handler (which is derived ultimately from `downshift`). 

This PR fixes #30204.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
In the linked issue (#30204) the reproducible steps can also be used to test:

1. Open the storybook for this project: `npm run storybook:dev`
2. Go to the `CustomSelectControl > Default` component, click it and start typing `s` to get to the entry `Small`
3. Notice the storybook sidebar no longer closes due to the `s` keypress being cancelled.

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ --> **Unsure, I had to add eslint-disable-next-line jsx-a11y/no-noninteractive-element-interactions but I don't see any other way of capturing this event**
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->

opr committed:
Stop the keydown event from propagating in CustomSelect component

opr commented:
Hi folks - coming in with an update to provide context. Here's a real-world example of what this PR aims to solve, when a user has a WordPress site, connected via Jetpack, there is a sidebar that can be opened by pressing N.

When a customer is using WooCommerce Blocks, there are input fields for Country/State that use `CustomSelectControl`.

If the user tries to type into these boxes, for example 'United Kingdom', then the `N` key is hit, causing the sidebar to open.

here's a video to show it happening (note the inputs given while the select box is focused are `u -> n` and the sidebar appears on the right-hand side immediately after `n`.)

https://www.loom.com/share/2dbe6a589c4541599438782d066151a7

ntsekouras commented on packages/components/src/custom-select-control/index.js:
```diff
@@ -128,7 +128,14 @@ export default function CustomSelectControl( {
+			<ul
+				{ ...menuProps }
+				onKeyDown={ ( e ) => {
+					e.nativeEvent.stopImmediatePropagation();
```
Why you're using `nativeEvent` here and also `stopImmediatePropagation` instead of `stopPropagation`?

opr replied:
Hi thanks for the review. I'm not sure why I did it that way, it was quite a while ago and I can't think of any justification for it.

I've changed it now and it has stopped the issue of `Escape` propagating to parent listeners.

ntsekouras reviewed (commented):
Thanks for working on this @opr !

After reading a bit more the [issue in downshift ](https://github.com/downshift-js/downshift/issues/630)and related discussions there, this approach seems to be the suggested workaround. I'm wondering though why even with this code, the `Escape` still propagated.

I inserted a `CustomSelectControl` in a Modal [here, after the TextControl](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/template-part/edit/placeholder/patterns-setup.js#L74)
-- demo snippet:
```
<CustomSelectControl
	label="Font Size"
	options={ [
		{
			key: 'small',
			name: 'Small',
			style: { fontSize: '50%' },
		},
		{
			key: 'normal',
			name: 'Normal',
			style: { fontSize: '100%' },
		},
	] }
/>
```

Then:
1.  insert a `Template Part` and click `New template part`
2. open the select control
3. press `Escape`

I would expect the Modal to stay open and only the select to close, but this is not happening..

opr committed:
Stop the synthetic event propagating instead of native

This stops other listeners receiving this event, whereas stopping the native event still somehow allowed propagation.

opr commented:
Thanks @ntsekouras I've made a change to stop the synthetic event propagating and it has solved the issue you mentioned in your review.

ciampo commented on packages/components/src/custom-select-control/index.js:
```diff
@@ -141,7 +141,14 @@ export default function CustomSelectControl( {
+				{ ...menuProps }
+				onKeyDown={ ( e ) => {
+					e.stopPropagation();
+					menuProps.onKeyDown( e );
```
I'm not sure we're guaranteed that `onKeyDown` is always defined on `menuProps` — it's probably safer to use optional chaining?

```suggestion
					menuProps?.onKeyDown?.( e );
```

ciampo commented on packages/components/src/custom-select-control/index.js:
```diff
@@ -141,7 +141,14 @@ export default function CustomSelectControl( {
+			{ /* eslint-disable-next-line jsx-a11y/no-noninteractive-element-interactions */ }
+			<ul
+				{ ...menuProps }
+				onKeyDown={ ( e ) => {
```
Could we refactor this inline function into a `handleKeyDown` function? We could even use the `useCallback` hook to memoize (with `menuProps` as the only dependency)

opr replied:
Good suggestion, thanks!

ciampo reviewed (commented):
Hey @opr , thank you for working on this fix!

I've added a couple of inline comments. On top of that, I've got two additional suggestions:

- Could you add a new entry to the [CHANGELOG](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CHANGELOG.md) ?
- Could you add a unit test for this fix?

Thank you!

opr committed:
Move keyDown handler into useCallback and use optional chaining

opr committed:
Add unit tests for CustomSelectControl

opr committed:
Update changelog

opr commented:
Hey @ciampo thanks for the review and sorry for the delay! All done and ready for a re-review :)

opr commented on packages/components/src/custom-select-control/test/index.js:
```diff
@@ -0,0 +1,31 @@
+	it( 'Captures the keypress event and does not let it propagate', () => {
+		const wrapper = (
+			<div
+				role="none"
```
To stop eslint reporting an issue with the div having a `keyDown` listener but no role.

ciampo replied:
Sounds good! Could you add this as an inline comment too?

ciampo commented on packages/components/src/custom-select-control/test/index.js:
```diff
@@ -0,0 +1,31 @@
+import { CustomSelectControl } from '@wordpress/components';
+
+describe( 'CustomSelectControl', () => {
+	const onKeyDown = jest.fn();
```
It's probably better if the `onKeyDown` spy gets instantiated _inside_ the test, to avoid potential bugs in the future when adding more tests and forgetting to reset the spy in between tests. i.e.

```js

describe( 'CustomSelectControl', () => {
  it( 'Captures the keypress event and does not let it propagate', () => {
    const onKeyDown = jest.fn();
    // ...
```

opr replied:
Thanks, good point!

ciampo commented on packages/components/CHANGELOG.md:
```diff
@@ -4,7 +4,8 @@
 ### Bug Fix
 
--   Add missing styles to the `BaseControl.VisualLabel` component. ([#37747](https://github.com/WordPress/gutenberg/pull/37747))
+- Add missing styles to the `BaseControl.VisualLabel` component. ([#37747](https://github.com/WordPress/gutenberg/pull/37747))
+- Prevent keyDown events from propagating up in `CustomSelectControl` ([#30557](https://github.com/WordPress/gutenberg/pull/30557))
```
Let's match the spacing with the rest of the CHANGELOG entries:

```suggestion
-   Add missing styles to the `BaseControl.VisualLabel` component. ([#37747](https://github.com/WordPress/gutenberg/pull/37747))
-   Prevent keyDown events from propagating up in `CustomSelectControl` ([#30557](https://github.com/WordPress/gutenberg/pull/30557))
```

opr replied:
Thanks!

ciampo commented on packages/components/src/custom-select-control/test/index.js:
```diff
@@ -0,0 +1,31 @@
+				className="test-class-name"
+				onKeyDown={ onKeyDown }
+			>
+				<CustomSelectControl options={ [ 'a', 'b', 'c' ] } />
```
The `options` prop should be an array of objects with a `key` and a `name` properties

```js
// ...
const options = [
  {
    key: 'one',
    name: 'Option one',
  },
  {
    key: 'two',
    name: 'Option two',
  },
  {
    key: 'three',
    name: 'Option three',
  },
];

render(
  <div role="none" onKeyDown={ onKeyDown }>
    <CustomSelectControl options={ options } />
  </div>
);
// ...
```

ciampo commented on packages/components/src/custom-select-control/test/index.js:
```diff
@@ -0,0 +1,31 @@
+		const wrapper = (
+			<div
+				role="none"
+				className="test-class-name"
```
I believe this className can be removed?

ciampo commented on packages/components/src/custom-select-control/test/index.js:
```diff
@@ -0,0 +1,31 @@
+				<CustomSelectControl options={ [ 'a', 'b', 'c' ] } />
+			</div>
+		);
+		const { container } = render( wrapper );
```
We could simplify the code by rendering the markup directly inside the `render` function:

```js
const { container } = render(
  <div role="none" onKeyDown={ onKeyDown }>
    <CustomSelectControl options={ options } />
  </div>
);
```

Actually, since we don't need `container` (see next comment), this could be just written as

```js
render(
  <div role="none" onKeyDown={ onKeyDown }>
    <CustomSelectControl options={ options } />
  </div>
);
```

opr replied:
Thanks, this is great to know :)

ciampo commented on packages/components/src/custom-select-control/test/index.js:
```diff
@@ -0,0 +1,31 @@
+			</div>
+		);
+		const { container } = render( wrapper );
+		fireEvent.keyDown(
+			container.getElementsByClassName(
+				'components-custom-select-control__menu'
+			)[ 0 ]
+		);
```
Using `getElementsByClassName` is not ideal, since we're [testing the implementation details](https://kentcdodds.com/blog/testing-implementation-details) of this component. Instead, we should write our tests from the point of view of the user, relying more on the accessible roles and the content that gets rendered — i.e:

1. open the select by clicking the toggle button
2. look for the element with role listbox (i.e. the equivalent of a `select`)
3. fire the event on that element

It is also a good practice to use the `screen` object instead of calling methods on the `container` (see [link](https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#not-using-screen))

This would roughly translate to this code:

```js
import { render, fireEvent, screen } from '@testing-library/react';

// ...

const toggleButton = screen.getByRole( 'button' );
fireEvent.click( toggleButton );

const customSelect = screen.getByRole( 'listbox' );
fireEvent.keyDown( customSelectControl );
```

opr replied:
Nice, thanks a lot for helping me out here! 👍🏼

ciampo commented on packages/components/src/custom-select-control/test/index.js:
```diff
@@ -0,0 +1,31 @@
+				'components-custom-select-control__menu'
+			)[ 0 ]
+		);
+		expect( onKeyDown.mock.calls.length ).toBe( 0 );
```
We could use the higher-level `toHaveBeenCalledTimes` [assertion](https://jestjs.io/docs/expect#tohavebeencalledtimesnumber):

```suggestion
		expect( onKeyDown ).toHaveBeenCalledTimes( 0 );
```

opr replied:
Should have known this 🤦🏼  cheers.

ciampo commented on packages/components/src/custom-select-control/test/index.js:
```diff
@@ -0,0 +1,31 @@
+ * External dependencies
+ */
+import { render, fireEvent } from '@testing-library/react';
+/**
```
Let's add an empty line here

```suggestion

/**
```

ciampo reviewed (commented):
Hey @opr , thank you for addressing the initial round of feedback!

I've left some additional feedback to align the unit tests with the most common practices used in the components package.

opr committed:
Add inline comment about role and remove unneeded class name

opr committed:
Move mocked function into test

opr committed:
Add correct spacing to changelog

opr committed:
Add options as an array of key value pairs

opr committed:
Use accessible roles instead of looking for elements by className

ciampo reviewed (approved):
Thank you @opr for swiftly addressing the feedback! Code changes LGTM 🚀 

The e2e test failures don't seem related, maybe rebasing would solve them?

opr commented:
> Thank you @opr for swiftly addressing the feedback! Code changes LGTM 🚀
> 
> The e2e test failures don't seem related, maybe rebasing would solve them?

Rebasing worked, thanks @ciampo
